nixos/nginx: first-class PROXY protocol support

RaitoBezarius opened the pull request:
###### Description of changes

This PR enables you to use PROXY protocol with maximum flexibility, as shown in the test.

An example of use-case is when you want to offer your legacy IP clients a way to connect to your IPv6-only server, but you want to have proper logs, so you will put up some TLS-SNI thing above NGINX (can be ngx_stream itself.)
While preventing anyone to spoof their IP address.

See https://www.mythic-beasts.com/support/topics/proxy for a compelling introduction of why this is interesting.

Please read the commit message for a summary.

I am running this PR in production on multiple servers at the moment to smoke out any problem.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented on nixos/tests/nginx-proxyprotocol/_.test.nix.cert.pem:
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
We can generate some certificates on the fly and don't need to commit them.

RaitoBezarius replied:
I already reused ACME machinery but it only provides for a domain and not a wildcard, so I just copied the stuff over there and run the scripts.

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -27,12 +27,7 @@ with lib;
-        ssl  = mkOption { type = bool; description = lib.mdDoc "Enable SSL.";  default = false; };
-        extraParameters = mkOption { type = listOf str; description = lib.mdDoc "Extra parameters of this listen directive."; default = []; example = [ "backlog=1024" "deferred" ]; };
-      }; });
+      type = with types; listOf (submodule (import ./listen-options.nix { inherit lib; }));
```
Can we just keep this inlined but well formated?

RaitoBezarius replied:
Yes, no problem.

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -297,24 +297,43 @@ let
             else defaultListen;
 
-        listenString = { addr, port, ssl, extraParameters ? [], ... }:
+        listenString = { addr, port, ssl, proxyProtocol ? false, extraParameters ? [], ... }:
```
Maybe we should go an approach like https://github.com/SuperSandro2000/nixpkgs/commit/374662311cd1dd01c12d278d69ca7e68124ed56f to make it easier to extend

RaitoBezarius replied:
Hm, but I'm not sure I want to apply extra parameters to all listen directives at the same time. Most of the time, I want some listen to be proxy_protocol and others not.

Adding `proxyProtocol` might be dropped because we can rely on `extraParameters` as you say.

SuperSandro2000 replied:
Yeah, you are right. proxy-connect you usually only want on one listen.

SuperSandro2000 reviewed (commented):
Don't we need some assertions around proxyProtocol? I vaguely remember that it disabled other protocols.

SuperSandro2000 commented on pkgs/servers/http/nginx/0001-Support-mixed-addr-configuration-for-PROXY-protocol-.patch:
```diff
@@ -0,0 +1,58 @@
+Subject: [PATCH] Support mixed addr configuration for PROXY protocol in HTTP
+ realip module
+
+This ensures that under `real_ip_header proxy_protocol`, if a request is
+received on a `listen` block which do not contain `proxy_protocol`, it
+is not rejected by NGINX.
```
I didn't think this through yet but how dangerous would this be with the following headers recommended by upstreams doc?
```
set_real_ip_from ${somePrivateIpv4};
set_real_ip_from ${somePrivateIpv6};

real_ip_header proxy_protocol;

proxy_set_header X-Real-IP       $proxy_protocol_addr;
proxy_set_header X-Forwarded-For $proxy_protocol_addr;
```

RaitoBezarius replied:
Let's try it! I will add something to the tests with that and let's see if we can create a situation where we spoof IPs.

RaitoBezarius replied:
Thinking out loud and trying some stuff:

- Without this patch, if you listen with proxy protocol, you can still be spoofed if you don't firewall properly the entering packets, and you trust blindly `$proxy_protocol_addr`
- `set_real_ip_from` is only a simple filtering control solution which aims to replace automatically `$remote_addr` and `$remote_port` out of the box based on `real_ip_header` directive.
- Upstream advises to use `set_real_ip_from` **OR** `proxy_set_header X-Real-IP ...`, etc. But upstream is a bit confusing in how they advise to do it.

In your scenario, I indeed can control `X-Real-IP` and `X-Forwarded-For` even under `set_real_ip_from` which is not here to prevent that.

My patch BTW under no proxy protocol does:
`            ngx_http_set_ctx(r, NULL, ngx_http_realip_module);`

which clears all the variables sets by NGINX and fallback to the usual ones.

Under proxy protocol, you have something called `    if (ngx_http_get_forwarded_addr(r, &addr, xfwd, value, rlcf->from,
                                    rlcf->recursive)
        != NGX_DECLINED)` which will be in charge of figuring out if that original IP was allowed to provide for a client IP with respect to `real_ip_recursive on/off` too.

So I think it's fine, except we need to improve the documentation on how to do all of this securely and why this is also important to firewall your flows to the PROXY-enabled listener if you are worried. :)

RaitoBezarius replied:
(I have a small test with `curl --haproxy-protocol` that I'm trying to make work to spoof some IP and show that my theory is right.)

RaitoBezarius replied:
(this is BTW further reinforced by guides like https://www.mythic-beasts.com/support/topics/proxy)

RaitoBezarius replied:
So… I dropped the patch as it is working without it :). I misunderstood everything about how NGINX worked **and** I was not able to reproduce my initial issues without the patch.

I tried to cook up some attack tests but `curl` is not really good for this usecase, I reached the author to ask if he was okay with more flexibility on this which would open up to attack testing, but this is not ready for now.

I put it as a FIXME and will revisit it in the next months, in the meantime, I will just do basic sanity checks by hand on deployed servers.

RaitoBezarius committed:
nixos/nginx: first-class PROXY protocol support

PROXY protocol is a convenient way to carry information about the
originating address/port of a TCP connection across multiple layers of
proxies/NAT, etc.

Currently, it is possible to make use of it in NGINX's NixOS module, but
is painful when we want to enable it "globally".
Technically, this is achieved by reworking the defaultListen options and
the objective is to have a coherent way to specify default listeners in
the current API design.
See `mkDefaultListenVhost` and `defaultListen` for the details.

It adds a safeguard against running a NGINX with no HTTP listeners (e.g.
only PROXY listeners) while asking for ACME certificates over HTTP-01.

An interesting usecase of PROXY protocol is to enable seamless IPv4 to
IPv6 proxy with origin IPv4 address for IPv6-only NGINX servers, it is
demonstrated how to achieve this in the tests, using sniproxy.

Finally, the tests covers:

- NGINX `defaultListen` mechanisms are not broken by these changes;
- NGINX PROXY protocol listeners are working in a final usecase
  (sniproxy);
- uses snakeoil TLS certs from ACME setup with wildcard certificates;

In the future, it is desirable to spoof-attack NGINX in this scenario to
ascertain that `set_real_ip_from` and all the layers are working as
intended and preventing any user from setting their origin IP address to
any arbitrary, opening up the NixOS module to bad™ vulnerabilities.

For now, it is quite hard to achieve while being minimalistic about the
tests dependencies.

RaitoBezarius commented:
> Don't we need some assertions around proxyProtocol? I vaguely remember that it disabled other protocols.

PROXY protocol disable normal HTTP requests, the only broken scenario I see is ACME retrieval of certificates.

If you have only PROXY listeners and you ask for a ACME certificate, it will fail.
Therefore, we should have an assert error for this kind of stuff and I think that's pretty much it?

RaitoBezarius commented:
I guess I should rebase on a recent revision of master, I'm running all the NixOS tests for NGINX at the moment, once this is good, I will rebase and it will be review time :).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1968

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -174,6 +174,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `services.mastodon` gained a tootctl wrapped named `mastodon-tootctl` similar to `nextcloud-occ` which can be executed from any user and switches to the configured mastodon user with sudo and sources the environment variables.
 
+- `services.nginx` gained a `defaultListen` option at server-level with support for PROXY protocol listeners, also `proxyProtocol` is now exposed in `services.nginx.virtualHosts.<name>.listen` option. It is now possible to run PROXY listeners and non-PROXY listeners at a server-level, see [#213510](https://github.com/NixOS/nixpkgs/pull/213510/) for more details.
```
Can you please copy the relevant part from your comment into here?

RaitoBezarius replied:
"the relevant part" -> what do you mean exactly?

SuperSandro2000 replied:
I am not sure but it would be nice if people while reading the changelog don't need to trace this PR and filter what is relevant. :)

RaitoBezarius replied:
Got it

RaitoBezarius replied:
good point

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -290,24 +290,43 @@ let
+          if cfg.defaultListen != [] then defaultListen1
+            # Cleanup nulls which will mess up with //.
+            # TODO: is there a better way to achieve this? i.e. mergeButIgnoreNullPlease?
+            (map (listenLine: filterAttrs (_: v: !(isNull v)) listenLine) cfg.defaultListen)
```
```suggestion
            (map (listenLine: filterAttrs (_: v: (v != null)) listenLine) cfg.defaultListen)
```

SuperSandro2000 commented on nixos/tests/nginx-proxyprotocol/generate-certs.nix:
```diff
@@ -0,0 +1,34 @@
+in mkDerivation {
+  name = "test-certs";
+  buildInputs = [ (minica.overrideAttrs (old: {
+    prePatch = ''
```
```suggestion
    postPatch = ''
```

RaitoBezarius replied:
Are you certain of this suggestion? This is what has been written in the original ACME files.

RaitoBezarius replied:
I applied it, hopefully, it should not break but it creates divergence from ACME original code.

SuperSandro2000 commented on nixos/tests/nginx-proxyprotocol/generate-certs.nix:
```diff
@@ -0,0 +1,34 @@
+  conf = import ./snakeoil-certs.nix;
+  domain = conf.domain;
+  domainSanitized = pkgs.lib.replaceStrings ["*"] ["_"] domain;
+in mkDerivation {
+  name = "test-certs";
```
We should probably use runCommand here

SuperSandro2000 commented on nixos/tests/nginx-proxyprotocol/default.nix:
```diff
@@ -0,0 +1,139 @@
+        '';
+      services.nginx = {
+        enable = true;
+        defaultListen = [ { addr = "127.0.0.1"; proxyProtocol = true; ssl = true; } { addr = "127.0.0.2"; } { addr = "127.0.0.3"; ssl = false; } { addr = "127.0.0.4"; ssl = false; proxyProtocol = true; }];
```
```suggestion
        defaultListen = [
          { addr = "127.0.0.1"; proxyProtocol = true; ssl = true; }
          { addr = "127.0.0.2"; }
          { addr = "127.0.0.3"; ssl = false; }
          { addr = "127.0.0.4"; ssl = false; proxyProtocol = true; }
        ];
```

SuperSandro2000 commented on nixos/tests/nginx-proxyprotocol/default.nix:
```diff
@@ -0,0 +1,139 @@
+        127.0.0.4 unsecure-nossl.test.nix
+        127.0.0.2 direct-noproxy.test.nix
+        127.0.0.1 direct-proxy.test.nix
+        '';
```
```suggestion
      '';
```

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -499,13 +519,31 @@ in
         example = literalExpression ''[ "10.0.0.12" "[2002:a00:1::]" ]'';
         description = lib.mdDoc ''
           If vhosts do not specify listenAddresses, use these addresses by default.
+          This is akin to writing `defaultListen = [ { addr = "0.0.0.0" } ... ]`.
```
```suggestion
          This is akin to writing `defaultListen = [ { addr = "0.0.0.0" } ]`.
```

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -499,13 +519,31 @@ in
       };
 
+      defaultListen = mkOption {
+        type = with types; listOf (submodule { options = {
+         addr = mkOption { type = str;  description = lib.mdDoc "IP address.";  };
+         port = mkOption { type = nullOr port;  description = lib.mdDoc "Port number."; default = null; };
+         ssl  = mkOption { type = nullOr bool; description = lib.mdDoc "Enable SSL."; default = null; };
+         proxyProtocol = mkOption { type = nullOr bool; description = lib.mdDoc "Enable PROXY protocol."; default = null; };
+         extraParameters = mkOption { type = nullOr (listOf str); description = lib.mdDoc "Extra parameters of this listen directive."; default = null; example = [ "backlog=1024" "deferred" ]; };
+        }; });
```
```suggestion
        type = with types; listOf (submodule {
          options = {
            addr = mkOption {
              type = str;
              description = lib.mdDoc "IP address.";
            };
            port = mkOption {
              type = nullOr port;
              description = lib.mdDoc "Port number.";
              default = null;
            };
            ssl  = mkOption {
              type = nullOr bool;
              default = null;
              description = lib.mdDoc "Enable SSL.";
            };
            proxyProtocol = mkOption {
              type = nullOr bool;
              description = lib.mdDoc "Enable PROXY protocol.";
              default = null;
            };
            extraParameters = mkOption {
              type = nullOr (listOf str);
              description = lib.mdDoc "Extra parameters of this listen directive.";
              default = null;
              example = [ "backlog=1024" "deferred" ];
            };
          };
        });
```
proxyProtocoll and ssl are bool types, why not default them to false?

RaitoBezarius replied:
`ssl = null;` means that you allow whatever makes sense at generation time.
whereas `ssl = false;` no SSL no matter what or `ssl = true;` SSL only

For `proxyProtocol`, it could indeed makes sense to have it as a boolean defaulting to false. Checking if this is possible.

RaitoBezarius replied:
Addressed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2089

RaitoBezarius commented:
Friendly ping @fpletz @ajs124 @emilazy @lukegb :).

RaitoBezarius commented:
We missed the deadline for a change in 23.05, I wish we can merge this post-branch off though.

RaitoBezarius commented:
We branched-off, I suggest to merge this to have plenty of time to deal with breakages @fpletz @ajs124 @emilazy @lukegb ? :)

RaitoBezarius commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -534,13 +552,53 @@ in
+            extraParameters = mkOption {
+              type = nullOr (listOf str);
+              description = lib.mdDoc "Extra parameters of this listen directive.";
+              default = null;
```
hmmm right!

RaitoBezarius commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -534,13 +552,53 @@ in
+          };
+        });
+        default = [];
+        example = literalExpression ''[ { addr = "10.0.0.12"; proxyProtocol = true; ssl = true; } { addr = "0.0.0.0"; } { addr = "[::0]"; } ]'';
```
good point

RaitoBezarius commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -304,36 +304,54 @@ let
         hasSSL = onlySSL || vhost.addSSL || vhost.forceSSL;
 
+        # First evaluation of defaultListen based on a set of listen lines.
+        defaultListen1 = listenLines:
```
`mkDefaultListenVhost` then? this is really some impl detail, I don't want to give too much meaning to it.

SuperSandro2000 replied:
lets use that one, better than ``defaultListen1``

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -1061,6 +1119,32 @@ in
+        # this operation.
+        # Options' priorities are the following at the moment:
+        # listen (vhost) > defaultListen (server) > listenAddresses (vhost) > defaultListenAddresses (server)
+        assertion =
```
Linear sounds good enough :)

RaitoBezarius commented:
@ofborg test nginx-proxyprotocol
